R/data.table.R

```diff
@@ -2415,6 +2415,9 @@ split.data.table = function(x, f, drop = FALSE, by, sorted = FALSE, keep.by = TR
     if (!missing(by))
       stopf("passing 'f' argument together with 'by' is not allowed, use 'by' when split by column in data.table and 'f' when split by external factor")
     # same as split.data.frame - handling all exceptions, factor orders etc, in a single stream of processing was a nightmare in factor and drop consistency
+    # evaluate formula #5392
+    if (inherits(f, "formula"))
+        f <- eval(attr(stats::terms.formula(f), "variables"), x, environment(f))
```

Comment:
`terms()` suffices right? because we already checked `inherits(f, "formula")`.

In fact, we should prefer it, suppose `class(f) = c("formulaExtension", "formula")`, and `terms.formulaExtension` is a better method.

Reply:
Good point, but `terms.formulaExtension` could also be a "worse" method which does not work for our case? Moreover, using `terms.formula` saves the dispatch of `terms.default`.

Reply:
then author should make it work better :)

dispatch is done once, overhead should be trivial. compared to allocating output elements & executing split. terms() is the better design IMO.